File: crates/apollo_network/src/bin/broadcast_network_stress_test_node/run/cluster_start.py
import argparse
import os
from time import sleep
import json
from cluster_port_forward_prometheus import cluster_port_forward_prometheus
from utils import (
    make_timestamp,
    run_cmd,
    pr,
    connect_to_cluster,
    project_root,
)
from yaml_maker import (
    get_prometheus_yaml_file,
    get_prometheus_deployment_yaml_file,
    get_prometheus_service_yaml_file,
    get_prometheus_headless_service_yaml_file,
    get_network_stress_test_deployment_yaml_file,
    get_network_stress_test_headless_service_yaml_file,
    get_namespace_deletion_job_yaml_file,
    get_namespace_deleter_rbac_yaml_file,
)
from args import add_shared_args_to_parser
from cluster_stop import (
    cluster_working_directory,
    broadcast_network_stress_test_deployment_file_name,
    stop_last_cluster_run,
)


def login_to_docker_registry():
    run_cmd("gcloud auth configure-docker us-central1-docker.pkg.dev")


def make_image_tag(timestamp: str) -> str:
    return f"us-central1-docker.pkg.dev/starkware-dev/sequencer/broadcast-network-stress-test-node:{timestamp}"


def build_image(image_tag: str):
    dockerfile_path = os.path.abspath("Dockerfile")
    run_cmd(f"docker build -t {image_tag} -f {dockerfile_path} {project_root()}")


def upload_image_to_registry(image_tag: str):
    run_cmd(
        f"docker push {image_tag}",
        hint="Make sure you are logged in to the Docker registry. If so, contact the dev team to resolve any issues (maybe a permissions issue).",
    )


def write_deployment_file(deployment_data: dict):
    with open(broadcast_network_stress_test_deployment_file_name, "w") as f:
        json.dump(deployment_data, f, indent=4)


def write_yaml_file(file_name: str, file_content: str):
    with open(os.path.join(cluster_working_directory, file_name), "w") as f:
        f.write(file_content)


def write_yaml_files(
    image_tag: str,
    args: argparse.Namespace,
    namespace_name: str,
    delay_seconds: int,
) -> list[str]:
    num_nodes = args.num_nodes
    files = {
        "broadcast-network-stress-test-deployment.yaml": get_network_stress_test_deployment_yaml_file(
            image_tag, args=args
        ),
        "broadcast-network-stress-test-headless-service.yaml": get_network_stress_test_headless_service_yaml_file(),
        "prometheus-config.yaml": get_prometheus_yaml_file(num_nodes),
        "prometheus-statefulset.yaml": get_prometheus_deployment_yaml_file(),
        "prometheus-service.yaml": get_prometheus_service_yaml_file(),
        "prometheus-headless-service.yaml": get_prometheus_headless_service_yaml_file(),
        "namespace-deleter-rbac.yaml": get_namespace_deleter_rbac_yaml_file(
            namespace_name
        ),
        "namespace-deletion-job.yaml": get_namespace_deletion_job_yaml_file(
            namespace_name, delay_seconds
        ),
    }

    for file_name, file_content in files.items():
        write_yaml_file(file_name, file_content)
    return list(files.keys())


class ExperimentRunner:
    def __enter__(self):
        self.timestamp = make_timestamp()
        run_cmd(f"mkdir -p {cluster_working_directory}")
        self.deployment_file = {"cluster_working_directory": cluster_working_directory}
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        write_deployment_file(self.deployment_file)

    def create_namespace(self, name_space_name: str):
        pr(f"Creating namespace {name_space_name}")
        self.deployment_file["namespace"] = name_space_name
        write_deployment_file(self.deployment_file)
        run_cmd(
            f"kubectl create namespace {name_space_name}",
        )

    def deploy_yaml_files(self, name_space_name: str):
        for file_name in self.deployment_file["yaml_files"]:
            pr(f"Deploying {file_name} to cluster")
            file_path = os.path.join(cluster_working_directory, file_name)

            # RBAC resources and namespace deletion job go to default namespace
            if file_name in [
                "namespace-deleter-rbac.yaml",
                "namespace-deletion-job.yaml",
            ]:
                run_cmd(
                    f"kubectl apply --wait -f {file_path} -n default",
                )
            else:
                run_cmd(
                    f"kubectl apply --wait -f {file_path} -n {name_space_name}",
                )

    def run_experiment(self, args: argparse.Namespace):
        pr(str(args))
        self.deployment_file["args"] = vars(args)
        image_tag = args.image if args.image else make_image_tag(self.timestamp)
        pr(f"timestamp: {self.timestamp}")
        self.deployment_file["timestamp"] = self.timestamp

        if args.image:
            self.deployment_file["was_image_built"] = False
        else:
            pr("Building image")
            build_image(image_tag)
            self.deployment_file["was_image_built"] = True
        pr(f"Image tag: {image_tag}")
        self.deployment_file["image_tag"] = image_tag
        run_cmd(
            f"docker image inspect {image_tag} > /dev/null",
            hint="Make sure the image exists before proceeding.",
        )

        connect_to_cluster()
        login_to_docker_registry()
        upload_image_to_registry(image_tag=image_tag)

        namespace_name = f"broadcast-network-stress-test-{self.timestamp}"
        delay_seconds = args.timeout + 300
        self.deployment_file["delay_seconds"] = delay_seconds

        self.create_namespace(namespace_name)
        file_names = write_yaml_files(
            image_tag,
            args=args,
            namespace_name=namespace_name,
            delay_seconds=delay_seconds,
        )
        self.deployment_file["yaml_files"] = file_names
        self.deploy_yaml_files(namespace_name)

        sleep(10)

        run_cmd(
            f"kubectl get pods -n {namespace_name}", hint="Check if pods are running"
        )
        pr("Prometheus deployment complete!")
        pr("To access Prometheus, run: python cluster_port_forward_prometheus.py")
        pr(
            f"Deployment files saved to: `{broadcast_network_stress_test_deployment_file_name}`"
        )

        # Print namespace deletion info
        pr(
            f"Namespace '{namespace_name}' will be automatically deleted after {self.deployment_file['delay_seconds']} seconds ({args.timeout} + 300)"
        )
        pr("This includes all resources within the namespace.")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--image",
        help="Previously built image tag to use instead of re-building the docker image.",
        type=str,
        default=None,
    )
    parser.add_argument(
        "--latency",
        help="Min latency to use when gating the network in milliseconds.",
        type=int,
        default=None,
    )
    parser.add_argument(
        "--throughput",
        help="Max throughput to use when gating the network in KB/s.",
        type=int,
        default=None,
    )
    parser.add_argument(
        "--dedicated-node-pool",
        help="Whether to run the pods on a dedicated node pool or not",
        action="store_true",
        default=False,
    )
    parser.add_argument(
        "--node-pool-name",
        help="Name of the dedicated node pool to use (only used if --dedicated-node-pool is set)",
        type=str,
        default="andrew",
    )
    parser.add_argument(
        "--node-pool-role",
        help="Role selector for the dedicated node pool (only used if --dedicated-node-pool is set)",
        type=str,
        default="andrew",
    )
    parser.add_argument(
        "--cpu-requests",
        help="CPU requests for each network stress test pod (in Kubernetes format, e.g., '1000m' for 1 core)",
        type=str,
        default="975m",  # running on machine with 3.92 CPU allocatable 3.9 / 4
    )
    parser.add_argument(
        "--memory-requests",
        help="Memory requests for each network stress test pod (in Kubernetes format, e.g., '1Gi' for 1 GiB)",
        type=str,
        default="1Gi",
    )
    parser.add_argument(
        "--cpu-limits",
        help="CPU limit for each network stress test pod (in Kubernetes format, e.g., '1000m' for 1 core)",
        type=str,
        default="975m",  # running on machine with 3.92 CPU allocatable 3.9 / 4
    )
    parser.add_argument(
        "--memory-limits",
        help="Memory limit for each network stress test pod (in Kubernetes format, e.g., '1Gi' for 1 GiB)",
        type=str,
        default="1Gi",
    )

    # parser.add_argument(
    #     "--timeout-seconds",
    #     help="Maximum duration for the stress test pods to run before automatic termination (seconds)",
    #     type=int,
    #     default=7200,
    # )

    add_shared_args_to_parser(parser=parser)
    args = parser.parse_args()

    if os.path.exists(broadcast_network_stress_test_deployment_file_name):
        x = input(
            "Deployment file already exists. Do you want to stop the last run? (y/N): "
        )
        if x.lower() == "y":
            pr("Stopping last cluster run...")
            stop_last_cluster_run()
        else:
            pr("Exiting without running the experiment.")
            return

    assert not os.path.exists(
        broadcast_network_stress_test_deployment_file_name
    ), "Deployment file already exists. Please run cluster_stop.py before running the experiment."

    with ExperimentRunner() as runner:
        runner.run_experiment(args)

    pr("Running cluster_port_forward_prometheus.py")

    cluster_port_forward_prometheus()


if __name__ == "__main__":
    main()

File: crates/apollo_network_benchmark/src/bin/apollo_network_benchmark_run/cluster_start.rs
use std::time::Duration;
use std::{fs, thread};

use anyhow::Context;
use clap::Parser;
use serde::Serialize;

use crate::args::{NetworkControls, ResourceLimits, SharedArgs, STRESS_TEST_NAME};
use crate::mod_utils::{
    build_docker_image,
    cluster_deployment_file_path,
    cluster_deployment_working_directory,
    connect_to_cluster,
    create_namespace,
    deploy_json_files_to_cluster,
    get_commit,
    login_to_docker_registry,
    make_cluster_image_tag,
    make_timestamp,
    run_cmd,
    run_with_deployment_guard,
    upload_image_to_registry,
    verify_docker_image_exists,
    write_deployment_file,
    write_json_file_to_working_dir,
};
use crate::pr;
use crate::yaml_maker::*;

#[derive(Parser, Debug, Serialize)]
pub struct ClusterStartArgs {
    #[command(flatten)]
    pub shared: SharedArgs,

    /// Previously built image tag to use instead of re-building the docker image
    #[arg(long)]
    pub image: Option<String>,

    #[command(flatten)]
    pub network_controls: NetworkControls,

    /// Role selector for the dedicated node pool
    #[arg(long, default_value = "apollo-network-benchmark")]
    pub node_pool_role: String,

    /// Node toleration in format key=value (e.g., "workload=apollo-network-benchmark")
    /// If set, adds a NoSchedule toleration to pods
    #[arg(long)]
    pub node_toleration: Option<String>,

    #[command(flatten)]
    pub resource_limits: ResourceLimits,

    /// Use production Docker build instead of development Docker build, defaults to faster
    /// development build if not specified
    #[arg(long, default_value = "false")]
    pub production_docker: bool,
}

fn write_json_files(
    image_tag: &str,
    args: &ClusterStartArgs,
    namespace_name: &str,
) -> anyhow::Result<Vec<String>> {
    let deployment_file_name = format!("{STRESS_TEST_NAME}-deployment.json");
    let headless_service_file_name = format!("{STRESS_TEST_NAME}-headless-service.json");
    let files = vec![
        (
            deployment_file_name.as_str(),
            get_network_stress_test_deployment_json_file(
                image_tag,
                &args.shared,
                &args.resource_limits,
                &args.network_controls,
                &args.node_pool_role,
                args.node_toleration.as_deref(),
            )?,
        ),
        (headless_service_file_name.as_str(), STRESS_TEST_HEADLESS_SERVICE_JSON.to_string()),
        ("prometheus-rbac.json", get_prometheus_rbac_json(namespace_name)?),
        ("prometheus-config.json", get_prometheus_json_file(args.shared.num_nodes)?),
        ("prometheus-statefulset.json", PROMETHEUS_DEPLOYMENT_JSON.to_string()),
        ("prometheus-service.json", PROMETHEUS_SERVICE_JSON.to_string()),
        ("prometheus-headless-service.json", PROMETHEUS_HEADLESS_SERVICE_JSON.to_string()),
        ("grafana-config.json", get_grafana_configmap_json_file()?),
        ("grafana-statefulset.json", GRAFANA_DEPLOYMENT_JSON.to_string()),
        ("grafana-service.json", GRAFANA_SERVICE_JSON.to_string()),
        ("grafana-headless-service.json", GRAFANA_HEADLESS_SERVICE_JSON.to_string()),
    ];

    let file_names: Vec<String> = files.iter().map(|(name, _)| name.to_string()).collect();

    for (file_name, file_content) in files {
        write_json_file_to_working_dir(file_name, &file_content)?;
    }

    Ok(file_names)
}

fn run_experiment(args: ClusterStartArgs) -> anyhow::Result<()> {
    pr!("{:?}", args);

    let timestamp = make_timestamp();
    let mut deployment_data = serde_json::json!({
        "cluster_working_directory": cluster_deployment_working_directory().display().to_string(),
        "timestamp": timestamp,
        "args": serde_json::to_value(&args).context("Failed to serialize args")?
    });

    let working_dir = cluster_deployment_working_directory();
    fs::create_dir_all(&working_dir).context("Failed to create working directory")?;

    let image_tag = if let Some(ref img) = args.image {
        deployment_data["was_image_built"] = serde_json::json!(false);
        img.clone()
    } else {
        pr!("Building image");
        let tag = make_cluster_image_tag(&timestamp);
        build_docker_image(&tag, !args.production_docker)?;
        deployment_data["commit"] = serde_json::json!(get_commit()?);
        deployment_data["was_image_built"] = serde_json::json!(true);
        tag
    };

    pr!("Image tag: {}", image_tag);
    deployment_data["image_tag"] = serde_json::json!(&image_tag);

    verify_docker_image_exists(&image_tag)?;

    connect_to_cluster()?;
    login_to_docker_registry()?;
    upload_image_to_registry(&image_tag)?;

    let namespace_name = format!("{STRESS_TEST_NAME}-{timestamp}");
    let delay_seconds = args.shared.user.timeout;
    deployment_data["delay_seconds"] = serde_json::json!(delay_seconds);
    deployment_data["namespace"] = serde_json::json!(&namespace_name);

    let deployment_file = cluster_deployment_file_path();
    write_deployment_file(&deployment_file, &deployment_data)?;

    create_namespace(&namespace_name)?;

    let file_names = write_json_files(&image_tag, &args, &namespace_name)?;
    deployment_data["json_files"] = serde_json::json!(file_names);

    write_deployment_file(&deployment_file, &deployment_data)?;

    deploy_json_files_to_cluster(&namespace_name, &file_names)?;

    thread::sleep(Duration::from_secs(10));

    run_cmd(
        &format!("kubectl get pods -n {}", namespace_name),
        "Check if pods are running",
        false,
    )?;

    pr!("Prometheus and Grafana deployment complete!");
    pr!("Deployment files saved to: `{}`", cluster_deployment_file_path().display());
    pr!("WARNING: Please don't forget to delete the namespace manually after the experiment is \
         complete !!!");
    pr!("");
    pr!("To set up port forwarding, run: cargo run --release --bin run -- cluster port-forward");

    Ok(())
}

pub fn run(args: ClusterStartArgs) -> anyhow::Result<()> {
    run_with_deployment_guard(
        &cluster_deployment_file_path(),
        || {
            pr!("Stopping last cluster run...");
            crate::cluster_stop::run()
        },
        || run_experiment(args),
    )
}
